### ISSUE TITLE
Token revocation

### ISSUE THREAD

**Issue by kbrownhbs** [clawrenceks/ReferenceTokenExchange#7]
Great work by the way.

I see that you are evicting JWTs from cache based on expiry.  But what if the reference token was revoked prior to expiry?  I don't see any place where revocation of reference token is checked.  Theoretically  one can continue using the JWT after the corresponding reference token had been revoked.  So ideally you would want to guard against that scenario and remove the JWT from cache?

**Comment by kbrownhbs** [comment]
Maybe one solution is to validate reference tokens after some threshold of time difference between current time and time since last validation.

For example, if the threshold is 5 minutes, then any cache hits that occur within 5 minutes of last reference token validation can be used without validation.  However, the first request received after the 5 minute threshold has been reached results in reference token validation.  If validation fails reference token/JWT pair is removed from cache.  If it succeeds then cache timestamp is updated to current time.

**Comment by kbrownhbs** [comment]
On second thought maybe just best to evict JWT from cache after 5 minutes.  Once evicted, discovery endpoint will be used to fetch new JWT.

**Comment by clawrenceks** [comment]
@kbrownhbs You are quite right, at this present time revoked reference tokens are not evicted from the cache, they remain there until they have expired. This is something that does need to be addressed, along with a second issue, which is allowing a single cache to be used across multiple instances of the service when it is scaled.

I will put some thought into this and share the proposed solution here. If you have any additional ideas then please feel free to share. I am not keen on arbitrarily removing all tokens from the cache after a predefined period of time however, this would defeat the purpose of the cache for me. I think I would favour an approach that would allow the cache to be notified when a revoked token should be removed from the cache.

**Comment by kbrownhbs** [comment]
Yes, it would probably be better to notify the cache to remove revoked or no longer valid tokens.

I do have some other ideas I will post as separate issues.
